Sources/SwiftFormatWhitespaceLinter/WhitespaceLinter.swift

```diff
@@ -0,0 +1,332 @@
+
+      // `userText` and `formattedText` should only differ in their whitespace characters.
+      if userNext.char != formNext.char {
+        fatalError("Characters do not match")
+      }
+
+      lastChar = userNext.char
+
+      compareWhitespace(
+        userOffset: userOffset,
+        isFirstCharacter: isFirstCharater,
+        userWs: userNext.whitespace,
+        formattedWs: formNext.whitespace
+      )
+
+      userOffset = userNext.offset + 1
+      formOffset = formNext.offset + 1
+      isFirstCharater = false
+    } while lastChar != nil
+  }
+
+  /// Compare the whitespace buffers between the user text and formatted text, and emit linter
+  /// errors accordingly.
+  ///
+  /// Note: properly formatted whitespace will always be some number of newlines characters,
+  /// followed by some number of spaces. e.g. "\n ", "\n\n  ", "\n", " ". The user's whitespace
+  /// could have spaces and newlines in any order. e.g. " \n ", "  \n", etc.
+  ///
+  /// - Parameters:
+  ///   - userOffset: The current non-whitespace character offset within the user text.
+  ///   - isFirstCharacter: Are we at the first character in the text?
+  ///   - userWs: The user leading whitespace buffer at the current character.
+  ///   - formattedWs: The formatted leading whitespace buffer at the current character.
+  func compareWhitespace(
+    userOffset: Int, isFirstCharacter: Bool, userWs: String, formattedWs: String
+  ) {
+    if userWs == formattedWs { return }
+
+    // e.g. "\n" -> ["", ""], and "" -> [""]
+    let userTokens = userWs.split(
+      separator: "\n", omittingEmptySubsequences: false).map(String.init)
+    let formTokens = formattedWs.split(
+      separator: "\n", omittingEmptySubsequences: false).map(String.init)
+
+    indentationError(
+      userOffset: userOffset,
+      isFirstCharacter: isFirstCharacter,
+      user: userTokens,
+      form: formTokens)
+
+    trailingWhitespaceError(userOffset: userOffset, user: userTokens, form: formTokens)
+
+    spacingError(
+      userOffset: userOffset,
+      isFirstCharacter: isFirstCharacter,
+      user: userTokens,
+      form: formTokens)
+
+    removeLineError(userOffset: userOffset, user: userTokens, form: formTokens)
+
+    addLinesError(userOffset: userOffset, user: userTokens, form: formTokens)
+  }
+
+  /// Compare user and formatted whitespace buffers, and check for indentation errors.
+  ///
+  /// Example:
+  ///   func myFun() {
+  ///   let a = 123  // Indentation error on this line
+  ///   }
+  ///
+  /// - Parameters:
+  ///   - userOffset: The current non-whitespace character offset within the user text.
+  ///   - isFirstCharacter: Are we at the first character in the text?
+  ///   - user: The tokenized user whitespace buffer.
+  ///   - form: The tokenized formatted whitespace buffer.
+  func indentationError(userOffset: Int, isFirstCharacter: Bool, user: [String], form: [String]) {
+    guard form.count > 1, user.count > 1 else {
```

Comment:
Nit: Use `&&` to join boolean conditions instead of the comma (the comma should be reserved to separate bindings that can't be used in a conjunction).

Reply:
Done.

It's actually a habit I picked up from MUMPS. You use commas to short-circuit `if` statements.

Reply:
> MUMPS

😱